lib/frrscript.c

```diff
@@ -183,61 +242,87 @@ void frrscript_register_type_codecs(struct frrscript_codec *codecs)
 		frrscript_register_type_codec(&codecs[i]);
 }
 
-struct frrscript *frrscript_load(const char *name,
-				 int (*load_cb)(struct frrscript *))
+struct frrscript *frrscript_new(const char *name)
 {
 	struct frrscript *fs = XCALLOC(MTYPE_SCRIPT, sizeof(struct frrscript));
 
 	fs->name = XSTRDUP(MTYPE_SCRIPT, name);
-	fs->L = luaL_newstate();
-	frrlua_export_logging(fs->L);
+	fs->lua_function_hash =
+		hash_create(lua_function_hash_key, lua_function_hash_cmp,
+			    "Lua function state hash");
+	return fs;
+}
+
+int frrscript_load(struct frrscript *fs, const char *function_name,
+		   int (*load_cb)(struct frrscript *))
+{
+
+	/* Set up the Lua script */
+	lua_State *L = luaL_newstate();
 
-	char fname[MAXPATHLEN * 2];
-	snprintf(fname, sizeof(fname), "%s/%s.lua", scriptdir, fs->name);
+	frrlua_export_logging(L);
 
-	int ret = luaL_loadfile(fs->L, fname);
+	char script_name[MAXPATHLEN * 2];
```

Comment:
Why `* 2`?
`MAXPATHLEN` is defined to the maximum length of a path on the filesystem, so if your path is any bigger than this, that's a hard error that needs to be handled

Reply:
Ah, I thought that MAXPATHLEN was the maximum path length that FRR takes, not the filesystem. This is from https://github.com/FRRouting/frr/commit/833c1f9fd14c4764b9f36f5e61c57130ff4f7118, but I suppose we shouldn't do that and let snprintf will help us cap script_name to MAXPATHLEN.

Reply:
The compiler expects "%s/%s.lua" to be put into a char [] of at least MAXPATHLEN + 5 since scriptdir is of MAXPATHLEN and +5 for the other characters and the script name. The `*2` made sure script_name was large enough. Should I stick with that or truncate the string somehow to exactly MAXPATHLEN?

Reply:
updated in https://github.com/FRRouting/frr/pull/8982/commits/99611df24ad8094a163db5cec5bb9bacd22e42b0. zlog error if path length is too long